python/ray/train/batch_predictor.py

```diff
@@ -187,14 +197,26 @@ def __init__(self):
 
             def __call__(self, batch):
                 if feature_columns:
-                    prediction_batch = batch[feature_columns]
+                    if batch_format == "numpy":
+                        prediction_batch = {
+                            k: v for k, v in batch.items() if k in feature_columns
+                        }
+                    else:
+                        prediction_batch = batch[feature_columns]
                 else:
                     prediction_batch = batch
                 prediction_output = self._predictor.predict(
                     prediction_batch, **predict_kwargs
                 )
                 if keep_columns:
-                    prediction_output[keep_columns] = batch[keep_columns]
+                    if batch_format == "numpy":
+                        for k, v in batch.items():
+                            if k in keep_columns:
+                                prediction_output[k] = v
+                    else:
+                        prediction_output[keep_columns] = batch[keep_columns]
+
+                # TODO(jiaodong): Add a dedicated casting exception type
                 return convert_batch_type_to_pandas(
```

Comment:
We could technically returned `prediction_output` uncasted, and Datasets will construct an Arrow Table from the ndarray or dict of ndarrays, but unlike the Pandas block representation, this would not be able to represent ragged tensors... I'm thinking that maybe the try-except fallback on the tensor extension casting makes sense here?